index.html

```diff
@@ -72,100 +74,160 @@
           data = JSON.parse(response);
 
-          if (preLength != data.length) {
+          if (preLength != response.length) {
+            preLength = response.length;
             // Clear markers
-            for (var i = 0; i < markers.length; i++) {
-              listeners[i].a.remove();
-              listeners[i].b.remove();
-              markers[i].setMap(null);
+            for (var pokehash in markers) {
+              if (!markers.hasOwnProperty(pokehash)) {
+                continue;
+              }
+              if (data.hasOwnProperty(pokehash) && hidden_layers[data[pokehash].id] != 1) {
+                continue;
+              }
+              var marker = markers[pokehash];
+              marker.listeners[0].remove();
+              marker.listeners[1].remove();
+              marker.setMap(null);
+              
+              delete markers[pokehash];
             }
-            listeners = [];
-            markers = [];
 
             // Add updated markers
-            for (var i = 0; i < data.length; i++) {
-              var p = data[i];
+            var numPkmnType = {};
+            for (var pokehash in data) {
+              if (!data.hasOwnProperty(pokehash)) {
+                // inherited, not interesting to us
+                continue;
+              }
+              var p = data[pokehash];
+              if (numPkmnType.hasOwnProperty(p.id)) {
+                numPkmnType[p.id] += 1;
+              } else {
+                numPkmnType[p.id] = 1;
+              }
               if (!updated) {
                 map.setZoom(16);
                 map.panTo({lat: p.lat, lng: p.lng});
                 updated = true;
               }
 
-              if(hidden_layers.indexOf(p.name) == -1){
+              if (layers.findIndex(function(v){ return v.id == p.id; }) == -1) {
+                console.log("Adding layer " + pkmn_label(p));
+                layers.push(p);
+                layers.sort(function(a, b){ return a.id - b.id; });
+                document.getElementById('pknum').innerHTML = " (" + (layers.length - 1) + ")";
+              }
+
+              if (hidden_layers[p.id] != 1) {
                 // Pokemon not hidden
-                if(layers.indexOf(p.name) == -1){
-                  layers.push(p.name);
+
+                p.timeleft = p.expiry - p.timestamp;
+                timeleft_strify = p.timeleft
+                p.timeleft_str = ''
+                if (timeleft_strify > 86400) {
+                    p.timeleft_str += Math.floor(timeleft_strify / 86400) + " days, ";
+                    timeleft_strify %= 86400;
+                }
+                if (timeleft_strify > 3600) {
+                    p.timeleft_str += Math.floor(timeleft_strify / 3600) + "h";
+                    timeleft_strify %= 3600;
+                }
+                if (timeleft_strify > 60) {
+                    p.timeleft_str += Math.floor(timeleft_strify / 60) + "m";
+                    timeleft_strify %= 60;
```

Comment:
Why you are doing all this mess when you can simply reset the layers array just like marker and listner in line 114 (that you removed)